### PR TITLE
[py] added decorator for handling deprecations.

### PR DESCRIPTION
**Pull request by sandeepsuryaprasad**
**Thanks for contributing to Selenium!**
**A PR well described will help maintainers to quickly review and merge it**

Before submitting your PR, please check our [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) guidelines.
Avoid large PRs, help reviewers by making them as simple and short as possible.

### Description
The motivation for this PR is https://github.com/SeleniumHQ/selenium/pull/12409
In this PR, I have created a decorator to handle deprecations in a much better and maintainable way.

### Motivation and Context
- I have created two decorators `deprecated_function` and `deprecated_attributes` in `selenium/deprecated.py`
- Any deprecated function would be decorated with `deprecated_function` decorator which takes care of logging warning message and from what version of selenium the function would be deprecated.
- Any deprecated instance variable would be decorated with `deprecated_attributes` decorator, which takes care of logging appropriate warning message for corresponding deprecated variable.
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.
- [] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by titusfortner** [comment]
This is a cool idea, but I'm not sure the advantage of this over a more straightforward method? Does it show up differently in the IDE with this annotation? Does the wrapper know the name of the function so we don't have to put it in the string?

File: py/selenium/deprecated.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+from warnings import warn
```
Comment (symonk) [inline py/selenium/deprecated.py:19]: Thanks @sandeepsuryaprasad will review more shortly, I put together a somewhat similar idea so i'm not too biased, open to having just have a reusable mechanism to handle it in general that works well.

one minor thing is new files should be prefixed with the licensing meta data

File: py/selenium/deprecated.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+removed_from_version = "4.12"
```
Comment (symonk) [inline py/selenium/deprecated.py]: I would rather we provided a \`version=\` flag maybe, what do you think? We are not great and getting deprecated things out atm, I think due to also not following semantic versioning makes it a tad harder (imo), but allowing us to define this vs hard coding would be useful

File: py/selenium/deprecated.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def deprecated_function(message):
+    """decorator to log deprecation warning messgaes for deprecated methods."""
+
+    def _deprecated_function(func):
+        def wrapper(*args, **kwargs):
```
Comment (symonk) [inline py/selenium/deprecated.py:29]: `@functools.wraps(func) here`

File: py/selenium/deprecated.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+            return func(*args, **kwargs)
+
+        return wrapper
+
+    return _deprecated_function
+
+
+def deprecated_attributes(*attrs, message):
+    def _deprecated_attributes(func):
+        def wrapper(*args, **kwargs):
```
Comment (symonk) [inline py/selenium/deprecated.py:41]: same here re: functools.wraps

File: py/selenium/deprecated.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            result = func(*args, **kwargs)
+            for attr in attrs:
+                _attr = getattr(args[0], attr, None)
+                if _attr:
+                    warn(f"'{message}': will be removed from {removed_from_version}", DeprecationWarning, stacklevel=2)
```
Comment (symonk) [inline py/selenium/deprecated.py]: does `stacklevel=2` play nice with a decorator (especially one with args?) in terms of the stack in the output? would need to double check

File: py/selenium/deprecated.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    return _deprecated_function
+
+
+def deprecated_attributes(*attrs, message):
+    def _deprecated_attributes(func):
+        def wrapper(*args, **kwargs):
+            result = func(*args, **kwargs)
+            for attr in attrs:
+                _attr = getattr(args[0], attr, None)
+                if _attr:
```
Comment (symonk) [inline py/selenium/deprecated.py]: not sure this work as you are expecting, e.g a deprecated attribute that been assigned `falsy`, e.g `log_path=False` would probably not warn here?

**Comment by symonk** [comment]
this stuff is quite trick with edge cases, for example a `@staticmethod` wouldn't have an instance to pass as `args[0]` if we ever needed to decorate it, is it simpler if we define a function and call that in places we are currently deprecating things? ie.

```python
from selenium.deprecations import deprecated

deprecated(message=..., version=4.12")  
```
Theres possibly alot of edge cases with the decorator approach.
Either way, i'll close my draft PR since we can colloborate as part of this work here.

File: py/selenium/deprecated.py (hunk 1 above)
Comment (sandeepsuryaprasad) [reply py/selenium/deprecated.py:19]: > Thanks @sandeepsuryaprasad will review more shortly, I put together a somewhat similar idea so i'm not too biased, open to having just have a reusable mechanism to handle it in general that works well.
> 
> one minor thing is new files should be prefixed with the licensing meta data

Yes, I have added license stuff.

File: py/selenium/deprecated.py (hunk 2 above)
Comment (sandeepsuryaprasad) [reply py/selenium/deprecated.py]: I have changed it to `version` now. The reason I had named it as `removed_from_version` is that, it should not be confused with current version of selenium. It is the version (future) from which something will be removed.

File: py/selenium/deprecated.py (hunk 3 above)
Comment (sandeepsuryaprasad) [reply py/selenium/deprecated.py:29]: Yes this is added.

File: py/selenium/deprecated.py (hunk 4 above)
Comment (sandeepsuryaprasad) [reply py/selenium/deprecated.py:41]: I have added here as well

File: py/selenium/deprecated.py (hunk 6 above)
Comment (sandeepsuryaprasad) [reply py/selenium/deprecated.py]: I have tweaked the logic. Now we can pass arbitrary keyword argument to the decorator where we can specify "should the user be warned if the deprecated attribute is falsy or should the user be warned if the he tries to pass some value to deprecated attribute which evaluates to truthy"

**Comment by sandeepsuryaprasad** [comment]
> This is a cool idea, but I'm not sure the advantage of this over a more straightforward method? Does it show up differently in the IDE with this annotation? Does the wrapper know the name of the function so we don't have to put it in the string?

Yes wrappers know the name of the function. `func.__name__` will give the name of the function which we have decorated.

**Comment by sandeepsuryaprasad** [comment]
> this stuff is quite trick with edge cases, for example a `@staticmethod` wouldn't have an instance to pass as `args[0]` if we ever needed to decorate it, is it simpler if we define a function and call that in places we are currently deprecating things? ie.
> 
> ```python
> from selenium.deprecations import deprecated
> 
> deprecated(message=..., version=4.12")  
> ```
> 
> Theres possibly alot of edge cases with the decorator approach. Either way, i'll close my draft PR since we can colloborate as part of this work here.

I have checked for `static` and `class` methods, `deprecated_function` decorator works fine in both cases. But if we want to make a function instead of decorator, that is also perfectly fine..

**Comment by sandeepsuryaprasad** [comment]
Getting `ModuleNotFound` error while trying to import the decorator from `deprecated.py` module. This is the file that is newly added to `selenium` folder.

**Comment by sandeepsuryaprasad** [comment]
Actually we don't have to write custom decorator. In `typing_extensions` module we do have `deprecated` decorator.
and in python 3.12 (yet to be released) there is a PEP to include this in `typing` module
https://peps.python.org/pep-0702/#example

**Comment by titusfortner** [comment]
aha, this is the PR I was looking for last week and couldn't find when I raised https://github.com/SeleniumHQ/selenium/pull/13402

I think we decided we want to use the annotation that is in `typing_extensions ` instead of this.

**Pull request closed without merging**